### ISSUE TITLE
Simplify `NativeFinalizer` to not need casts

### ISSUE THREAD

**Issue by Levi-Lesches** [dart-lang/sdk#53080]
This is based on just a few minutes of playing around so this might not work for reasons I'm not aware of, but I believe the `NativeFinalizer` class could use type arguments to avoid taking a `Void Function(Pointer<Void>)` (and avoid the `.cast()` as well).

```dart
/// Represents a `void nativeFinalizer(T* ptr)`
typedef MyNativeFinalizerFunction<T extends NativeType> = NativeFunction<Void Function(Pointer<T>)>;

/// `D` is the Dart (wrapper) type, `N` is the native type.
class MyFinalizer<D extends Finalizable, N extends NativeType> {
  final Pointer<MyNativeFinalizerFunction<N>> callback;
  MyFinalizer(this.callback);

  void attach(D obj, Pointer<N> ptr) { /* ... */ }
  void detach(D obj) { /* ... */ }
}
```
Would this be possible? I was able to swap my code's `NativeFinalizer` with `MyFinalizer` with no issues and was able to remove the `.cast`s from my code without issues as well. This way, existing code will keep working but new code can be more type-safe.

**Comment by dcharkes** [comment]
We have an `Allocator` abstraction that constructs `Pointer`s with the desired type argument. The `free` from the allocator takes `Pointer<dynamic>`. However, with the `NativeFinalizer`s the native function must take void.

Now we can't make the `NativeFunction` take `Pointer<NativeType>` because we would still require casts as the type argument is in a covariant position. However, we could reduce the number of casts by letting it take a type argument.

If a user binds to `free` with a `void*` directly, we have no casts. But if users have some helper struct, then actually it could be useful to have the finalizer function type and the helper struct type be matched here.

cc @mkustermann @mraleph @lrhn @liamappelbe @HosseinYousefi 

@Levi-Lesches Why would you also want a type argument `D` for the `Finalizable`? You can never get the set of objects out. Do you have situations where you have multiple different `Finalizable`s and you would like to prevent passing the wrong finalizable subtype to the wrong native finalizer? You could also achieve that by wrapping the native finalizer in another class I believe.

**Comment by Levi-Lesches** [comment]
> Now we can't make the `NativeFunction` take `Pointer<NativeType>` because we would still require casts as the type argument is in a covariant position. However, we could reduce the number of casts by letting it take a type argument.
> 
> If a user binds to `free` with a `void*` directly, we have no casts. But if users have some helper struct, then actually it could be useful to have the finalizer function type and the helper struct type be matched here.

Right, even if the finalizer callback and its value still require casting, it would be nice for the API to be type safe and do the casting internally.

> Why would you also want a type argument `D` for the `Finalizable`?

I don't know if it's necessary, so I included it to be safe. I was thinking: 
- To ensure that the `Finalizable` actually holds the pointer being freed. Explicitly declaring the type in the constructor would force you to pass in the right type in `attach`
- Ensuring that `attach(detach: x)` and `detach(x)` get the same object `x`. But they can be the same type and not the same object so that doesn't really work.
